### PR TITLE
fix: make WireLogger concurrency safe - WPB-11890

### PR DESCRIPTION
**Pull request by caldrian**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-11890" title="WPB-11890" target="_blank">TaskWPB-11890</a>  [iOS] Make `WireLogger` swift concurrency safe
  </td></table>

### Issue

This PR fixes some warnings regarding concurrency safety.

---

### Checklist

- [x] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [x] Description is filled and free of optional paragraphs.
- [x] Adds/updates automated tests.

---

### UI accessibility checklist

_If your PR includes UI changes, please utilize this checklist:_
- [x] Make sure you use the API for UI elements that support large fonts.
- [x] All colors are taken from WireDesign.ColorTheme or constructed using WireDesign.BaseColorPalette.
- [x] New UI elements have Accessibility strings for VoiceOver.

### CODE REVIEW

File: wire-ios-system/Source/Logging/SystemLogger.swift [hunk 1]
```diff
@@ -90,5 +92,3 @@ struct SystemLogger: LoggerProtocol {
         }
     }
 }
-
-private var loggers: [String: OSLog] = [:]
```
Comment (caldrian) [inline wire-ios-system/Source/Logging/SystemLogger.swift:94]: This array doesn't seem to be set or modified anywhere, so I deleted it.

File: wire-ios-system/Source/Logging/WireLogger.swift [hunk 2]
```diff
@@ -16,12 +16,18 @@
...
+    public static func initialize(loggers: [any LoggerProtocol]) {
+        guard provider == nil else {
+            assertionFailure("WireLogger.initialize called more than once")
+            return
+        }
+
+        provider = AggregatedLogger(loggers: loggers)
+    }
+
+    private static nonisolated(unsafe) var provider: (any LoggerProtocol)!
```
Comment (caldrian) [inline wire-ios-system/Source/Logging/WireLogger.swift]: The `provider` value is now injected via `initialize(loggers:)`.
This func should only be called once.
Still, I don't consider it necessary to use any custom logic to prevent. So I put a simple `guard` + `assertionFailure` there.

Currently this code is called by `WireAnalytics.setup()` which is called in
`UIApplicationDelegate.willFinishLaunchingWithOptions` and in the two app extensions.

File: wire-ios-system/Source/Logging/WireLogger.swift [hunk 3]
```diff
@@ -92,8 +98,4 @@ public struct WireLogger: LoggerProtocol {
     public static var logFiles: [URL] {
         provider.logFiles
     }
-
-    public static func addLogger(_ logger: LoggerProtocol) {
-        provider.addLogger(logger)
-    }
```
Comment (caldrian) [inline wire-ios-system/Source/Logging/WireLogger.swift]: By removing this the type can conform to `Sendable`.

File: wire-ios-system/Tests/ZMLogTests.swift [hunk 4]
```diff
@@ -362,6 +363,7 @@ extension ZMLogTests {
         Thread.sleep(forTimeInterval: 0.2)
     }
 
+    @MainActor
```
Comment (caldrian) [inline wire-ios-system/Tests/ZMLogTests.swift:366]: ... fixes warnings regarding `waitForExpectations`.

File: .github/pull_request_template.md [hunk 5]
```diff
@@ -11,7 +11,6 @@ _Optional: add details about technical approach, solutions etc._
 
 _Optional: reference dependencies to other pull requests etc._
 
-
```
Comment (caldrian) [inline .github/pull_request_template.md]: I've been deleting this line for every PR, so I decided to just do it once and for all here.
Please let me know if you prefer this to be reverted or done in another PR.

File: wire-ios/WireCommonComponents/Analytics/WireAnalytics.swift [hunk 6]
```diff
@@ -16,5 +16,31 @@
...
+import WireSystem
+
 /// Namespace for analytics tools.
-public enum WireAnalytics {}
+public enum WireAnalytics {
+
+    private static var isSetUp = false
+
+    public static func setup() {
+        guard !isSetUp else {
```
Comment (samwyndham) [inline wire-ios/WireCommonComponents/Analytics/WireAnalytics.swift:34]: I think we need to have some kind of lock here. I think we had issues without it in the notification service right @netbe ?

**Review by samwyndham — COMMENTED** [review]
Nice work. Much better than my proposal I think to introduce a global actor. Just left a comment about `isSetUp` and threading.

File: wire-ios/WireCommonComponents/Analytics/WireAnalytics.swift (hunk 6 above)
Comment (netbe) [reply wire-ios/WireCommonComponents/Analytics/WireAnalytics.swift:34]: yes the NSE can be reused multiple times, we should guarantee that the logger is added only once

Comment (netbe) [reply wire-ios/WireCommonComponents/Analytics/WireAnalytics.swift:34]: so it should be thread safe too

Comment (netbe) [reply wire-ios/WireCommonComponents/Analytics/WireAnalytics.swift:34]: I would keep the lock @caldrian unless there's something I am missing.

File: wire-ios-system/Source/Logging/CocoaLumberjackLogger.swift [hunk 7]
```diff
@@ -20,46 +20,46 @@ import CocoaLumberjackSwift
 import Foundation
 
 /// Logger to write logs to fileSystem via CocoaLumberjack
-final class CocoaLumberjackLogger: LoggerProtocol {
+public final class CocoaLumberjackLogger: LoggerProtocol {
```
Comment (netbe) [inline wire-ios-system/Source/Logging/CocoaLumberjackLogger.swift]: why do we need the classes to be public?

**Review by netbe — CHANGES_REQUESTED** [review]
I think we should keep the lock to avoid multiple initialization of the logger

File: wire-ios-system/Source/Logging/CocoaLumberjackLogger.swift (hunk 7 above)
Comment (caldrian) [reply wire-ios-system/Source/Logging/CocoaLumberjackLogger.swift]: I haven't moved the file but just used it from WireCommonComponents.
I'll give it a try to move the file and revert this.

File: wire-ios/WireCommonComponents/Analytics/WireAnalytics.swift (hunk 6 above)
Comment (caldrian) [reply wire-ios/WireCommonComponents/Analytics/WireAnalytics.swift:34]: Thanks, added a lock.

File: wire-ios-system/Source/Logging/WireLogger.swift [hunk 8]
```diff
@@ -34,46 +40,46 @@ public struct WireLogger: LoggerProtocol {
     // MARK: - LoggerProtocol
 
     public var logFiles: [URL] {
-        Self.provider.logFiles
+        Self.provider?.logFiles ?? []
```
Comment (caldrian) [inline wire-ios-system/Source/Logging/WireLogger.swift:43]: I made `provider` optional, because for unit tests the `initialize` method is not called.
Do you agree with this approach or do you suggest something different?

**Review by netbe — APPROVED** [review]

File: wire-ios/WireCommonComponents/Analytics/WireAnalytics.swift [hunk 9]
```diff
@@ -16,5 +16,39 @@
...
+    private static var isSetUp = false
+
+    public static func setup() {
+        // Adding a lock here since some app extension might execute this setup in the same process as the main app.
+        // https://stackoverflow.com/a/62674277
+        isSetUpLock.lock()
+        defer { isSetUpLock.unlock() }
+
+        guard !isSetUp else {
+            assertionFailure("WireAnalytics.setup() called more than once")
```
Comment (samwyndham) [inline wire-ios/WireCommonComponents/Analytics/WireAnalytics.swift:35]: I guess if https://stackoverflow.com/a/62674277 happens we might not want to assert here... It is normal behavior.

**Review by samwyndham — APPROVED** [review]

File: wire-ios/WireCommonComponents/Analytics/WireAnalytics.swift (hunk 9 above)
Comment (caldrian) [reply wire-ios/WireCommonComponents/Analytics/WireAnalytics.swift:35]: True, let's change it if it happens during debugging.
For release builds this statement will have no effect.

Comment (samwyndham) [reply wire-ios/WireCommonComponents/Analytics/WireAnalytics.swift:35]: FYI it happened to me in a debug build. I will remove the assertion.
